src/main/java/Duke.java

```diff
@@ -108,6 +116,91 @@ public static void main(String[] args) {
         }
     }
 
+    // Handlers
+
+    private static <T extends Task> void OnNewTaskAdded(T t) {
+        taskList.add(t);
+        writeToFile(t.PopulateSaveData() + System.lineSeparator());
+        printNewTask(t);
+    }
+
+    private static void OnTaskRemoved(int index) {
+        PrintWithIndent("Noted. I've removed this task: ");
+        PrintWithIndent("  " + taskList.get(index));
+        taskList.remove(index);
+        PrintWithIndent("Now you have " + taskList.size() + " tasks in the list.");
+        try {
+            List<String> originalContent = Files.readAllLines(SavePath);
+            originalContent.remove(index);
+            Files.write(SavePath, originalContent);
+        } catch (IOException e) {
+            PrintWithIndent("Error occurs when removing a task: " + e);
+        }
+    }
+
+    private static void OnTaskDone(int index) {
+        Task task = taskList.get(index);
+        task.isDone = true;
+        PrintWithIndent("Nice! I've marked this task as done: ");
+        PrintWithIndent("  " + task);
+        try {
+            List<String> originalContent = Files.readAllLines(SavePath);
+            originalContent.set(index, task.toString());
+            Files.write(SavePath, originalContent);
+        } catch (IOException e) {
+            PrintWithIndent("Save error (mark task): " + e);
+        }
+    }
+
+
+    // I/O methods
+
+    private static void InitializeSaveData() {
+        try {
+            Path path = Arrays.stream(SaveDirectory)
+                    .map(Paths::get)
+                    .reduce(Paths.get(""), Path::resolve)
+                    .resolve(SaveFileName);
+
+            if (!Files.exists(path)) {
+                Path p = Paths.get("");
+                for (String s : SaveDirectory) {
+                    p = p.resolve(s);
+                    if (!Files.exists(p))
+                        Files.createDirectory(p);
+                }
+
+                if (!Files.exists(path)) {
+                    Files.createFile(path);
+                }
+            }
+
+            SavePath = path;
+            List<String> fileContents = Files.readAllLines(path);
+            for(String s : fileContents) {
+                try {
+                    taskList.add(Task.LoadFromSaveData(s));
+                } catch (DukeException e) {
+                    PrintWithIndent(e.getMessage());
+                }
+            }
+        } catch (IOException e) {
+            PrintWithIndent("Initialize Error: Failed to create saving file.");
+        }
+    }
+
+    private static void writeToFile(String textToWrite) {
+        try {
+            FileWriter fileWriter = new FileWriter(SavePath.toFile(), true);
+            fileWriter.write(textToWrite);
+            fileWriter.close();
+        } catch (IOException e) {
+            PrintWithIndent("Error occurs when : " + e);
+        }
+    }
+
```

Comment:
This two methods could be written in a separate java file instead for more OOP, and both of them can also be simplified for ease of reading (in terms of implementation), and to reduce your class implementation length.